cfme/cloud/keypairs.py

```diff
@@ -119,24 +119,67 @@ def is_displayed(self):
     form = View.nested(KeyPairAddForm)
 
 
-class KeyPair(Taggable, SummaryMixin, Navigatable):
+class KeyPairCollection(Navigatable):
+    """ Collection object for the :py:class: `cfme.cloud.KeyPair`. """
+
+    def instantiate(self, name, provider, public_key=None):
+        return KeyPair(name, provider, public_key=public_key or "", collection=self)
+
+    def create(self, name, provider, public_key=None, cancel=False):
+        """Create new keyPair.
+
+        Args:
+            name (str): name of the KeyPair
+            public_key (str): RSA Key if present
+            provider (str): Cloud Provider
+            cancel (boolean): Cancel Keypair creation
+        """
+
+        view = navigate_to(self, 'Add')
+        changed = view.form.fill({'name': name,
+                                  'public_key': public_key,
+                                  'provider': provider.name
+                                  })
+        if cancel and not changed:
+            view.form.cancel.click()
+            flash_message = 'Add of new Key Pair was cancelled by the user'
+        else:
+            view.form.add.click()
+            flash_message = VersionPick({
+                Version.lowest(): 'Creating Key Pair {}'.format(name),
+                '5.8': 'Key Pair "{}" created'.format(name)}).pick(self.appliance.version)
+
+        # add/cancel should redirect, new view
+        view = self.create_view(KeyPairAllView)
+        # TODO BZ 1444520 causing ridiculous redirection times after submitting the form
+        wait_for(lambda: view.is_displayed, num_sec=120, delay=3,
+                 fail_func=view.flush_widget_cache, handle_exception=True)
+        view = self.create_view(KeyPairAllView)
+        assert view.is_displayed
+        view.entities.flash.assert_success_message(flash_message)
+        keypair = self.instantiate(name, provider, public_key=public_key)
+        return keypair
+
+
+class KeyPair(Navigatable):
     """ Automate Model page of KeyPairs
 
     Args:
         name: Name of Keypairs.
     """
     _param_name = "KeyPair"
 
-    def __init__(self, name, provider, public_key=None, appliance=None):
-        Navigatable.__init__(self, appliance=appliance)
+    def __init__(self, name, provider, public_key=None, appliance=None, collection=None):
+        self.collection = collection or KeyPairCollection(appliance=appliance)
+        Navigatable.__init__(self, appliance=self.collection.appliance)
         self.name = name
         self.provider = provider
-        self.public_key = public_key
+        self.public_key = public_key or ""
 
     def delete(self, cancel=False, wait=False):
         view = navigate_to(self, 'Details')
         view.toolbar.configuration.item_select('Remove this Key Pair',
-                                               handle_alert=not cancel)
+                                               handle_alert=True)
```

Comment:
**Required**
`cancel` arg is not actually used here to handle the alert, meaning if cancel is passed as True, this method will still delete the keypair.

Reply:
@mshriver do I need to add `not cancel` back to code ..?

Reply:
Looks like it, the differences for `cancel` look like they're still handled elsewhere.

Where I had to implement this, I wrote it with () so it was a bit more clear on reading:
`handle_alert=(not cancel)`

Reply:
I will update like this `handle_alert=(not cancel)`.